deadlines/src/main/java/com/palantir/deadlines/Deadlines.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * (c) Copyright 2025 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.deadlines;
+
+import com.palantir.deadlines.api.DeadlinesHttpHeaders;
+import com.palantir.tracing.TraceLocal;
+import java.time.Duration;
+import java.util.Optional;
+
+/**
+ * Utility methods for working with deadlines.
+ */
+public class Deadlines {
+
+    private static final TraceLocal<ProvidedDeadline> deadlineState = TraceLocal.of();
+
+    public static void putDeadline(Duration deadline) {
+        // store deadline state in a TraceLocal
+        // what do we do if there is an existing state stored in the TraceLocal already??
```

Comment:
We could do something like CloseableTracer where a closeable `DeadlineScope` is returned, which can reset the value after a block completes.

However, we'll have to deal with the issue of multiple threads in the same trace setting deadlines for specific code, which would work a great deal better as a [`ScopedValue`](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/ScopedValue.html), but that's still a preview in JDK-24: https://openjdk.org/jeps/487

Reply:
per our discussion offline, i think for a first-pass we should assume that a deadline value is write-once, read-many, and is only set once by the RPC stack reading a request header. That simplifies handling significantly, as we can assume multiple threads can only read the remaining deadline value, not create their own deadlines for internal work (though they may make outbound RPC calls with a new deadline set, which is fine)